### PR TITLE
Use RFC3339 in UTC as the default time format for log messages

### PR DESCRIPTION
**Pull request by efd6**
This ensures that times that are logged in messages can be deserialised by the encoding/json package irrespective of the timezone that the time is in when logged; currently if the time is not in UTC, the ISO8601 format that is used is not accepted. Also provide a non-UTC formatter that can be used where the timezone is important.

### CODE REVIEW

**Review by belimawr — COMMENTED** [review]
I'm just a bit confused why we need `RFC3339UTCTimeEncoder` and `RFC3339TimeEncoder` if only `RFC3339UTCTimeEncoder` is used directly and the only thing it does is to call `RFC3339TimeEncoder` with the time in UTC, which is the first thing `RFC3339TimeEncoder` does.

@efd6 could you clarify what is the effective difference between those two functions?

**Comment by efd6** [comment]
That is a bug.

**Review by belimawr — APPROVED** [review]

**Review by cmacknz — APPROVED** [review]

**Comment by efd6** [comment]
There are already tests for this.

**Comment by efd6** [comment]
That's specified [here](https://pkg.go.dev/time#Time.UnmarshalJSON).

**Comment by axw** [comment]
I think @kruskall is thinking of a test that the _default_ timestamp encoder is able to accept non-UTC timestamps, for this library.

Put another way: can we add a test that would fail without the change in this PR?

**Comment by axw** [comment]
@efd6 how about this?

```diff
diff --git a/logger_test.go b/logger_test.go
index 5f08f74..c72335c 100644
--- a/logger_test.go
+++ b/logger_test.go
@@ -77,10 +77,13 @@ func (tw *testOutput) validate(t *testing.T, keys ...string) {
 			case "string":
 				_, ok = val.(string)
 			case "datetime":
-				var s string
-				s, ok = val.(string)
-				if _, err := time.Parse("2006-01-02T15:04:05.000Z0700", s); err == nil {
-					ok = true
+				if s, strOK := val.(string); strOK {
+					if _, err := time.Parse("2006-01-02T15:04:05.000Z0700", s); err == nil {
+						var t time.Time
+						if err := t.UnmarshalText([]byte(s)); err == nil {
+							ok = true
+						}
+					}
 				}
 			case "integer":
 				// json.Unmarshal unmarshals into float64 instead of int
@@ -160,3 +163,29 @@ func TestECSZapLogger(t *testing.T) {
 		})
 	}
 }
+
+func TestECSZapLoggerTimezone(t *testing.T) {
+	out := testOutput{}
+
+	eucla, err := time.LoadLocation("Australia/Eucla")
+	require.NoError(t, err)
+
+	core := NewCore(NewDefaultEncoderConfig(), &out, zap.DebugLevel)
+	logger := zap.New(core, zap.WithClock(&locationClock{
+		Clock:    zapcore.DefaultClock,
+		location: eucla,
+	}))
+	defer logger.Sync()
+
+	logger.Info("hello from UTC+8:45")
+	out.validate(t, "@timestamp")
+}
+
+type locationClock struct {
+	zapcore.Clock
+	location *time.Location
+}
+
+func (c *locationClock) Now() time.Time {
+	return c.Clock.Now().In(c.location)
+}
```

**Comment by efd6** [comment]
The time validation in `validate` is not right; it assumes the incorrect ISO8601 format. I've gone with a behaviour-based test that looks to whether `json.Unmarshal` cope with it only.

The config test is enough to show that the change is required.

File: logger_test.go [hunk 1]
```diff
@@ -79,7 +79,13 @@ func (tw *testOutput) validate(t *testing.T, keys ...string) {
...
 				var s string
 				s, ok = val.(string)
-				if _, err := time.Parse("2006-01-02T15:04:05.000Z0700", s); err == nil {
+				if !ok {
+					break
+				}
+				var tv time.Time
+				// Depends on the agreement between Go quoting and JSON quoting
+				// for time values.
+				if err := json.Unmarshal([]byte(strconv.Quote(s)), &tv); err == nil {
```
Comment (axw) [inline logger_test.go:88]: We could alternatively use UnmarshalText here, and remove the need for quoting

**Review by axw — APPROVED** [review]
thanks for adding the test, looks good

File: logger_test.go (hunk 1 above)
Comment (axw) [reply logger_test.go:88]: no big deal, resolving so we can merge

**Pull request merged**